autogen/agentchat/contrib/nexusravenv2_local_function_calling.py

```diff
@@ -0,0 +1,162 @@
+import json
+import re
+from typing import Dict, Union, Tuple
+import autogen
+
+from typing_extensions import override
+
+""" This module contains the NexusFunctionCallingAssistant class, which is a subclass of the autogen.AssistantAgent class.
+    Its used specifically for calling functions locally using <a href="https://huggingface.co/TheBloke/NexusRaven-V2-13
+    The model can be imported from hugging face and the can use the standard function calling decorator to register the functions to be called.
+    If you enjoyed this module consider giving me a follow on github or buying me a coffee at https://www.buymeacoffee.com/gregnwosu
+    """
+
+
+def create_nexus_prompt_for_tool(tool: dict) -> str:
+    ''' takes a dictionary of form
+        {'type': 'function',
+         'function': {'description': 'This is a random word generator.',
+                      'name': 'random_word_generator',
+                      'parameters': {'type': 'object',
+                                     'properties': {'seed':   {'type': 'integer',
+                                                                'description': 'seed for initialising the random generator'},
+                                                     'prefix': {'type': 'string',
+                                                                'description': 'string to prefix random word'}},
+                                      'required': ['seed', 'prefix']}}}
+        and produces a string of the form:
+        
+        Function:
+        def random_word_generator(seed, pre):
+            """
+            This is a random word generator.
+
+            Args:
+            seed (integer): seed for initialising the random generator.
+            prefix (string): string to prefix random word.
+
+            Returns:
+            float: a random word
+    """
+    '''
+    function = tool['function']
+    name = function['name']
+    description = function['description']
+    parameters = function['parameters']
+    properties = parameters['properties']
+
+    args = ', '.join([f"{k}:{v['type']}" for k, v in properties.items()])
+    arg_types = ',\n'.join(
+        [f"{key} ({properties[key]['type']}): {properties[key]['description']}" for key in properties.keys()])
+    return f'''OPTION:\n<func_start>def {name}({args})<func_end>\n<docstring_start>\n"""\n{description.strip()}\n\nArgs:\n{arg_types}\n"""\n<docstring_end>'''
+
+
+def add_nexus_raven_prompts(func):
```

Comment:
Would you consider using a custom model client to handle the request and response transformation? Should be a better encapuslation of these transformation logics.

Reply:
I would need to understand how to create it. Would it be enough just to extend ConversableAgent and copy the _ generate, functions or would that be too much duplication  is there are more elegant way?

Reply:
im actually coming around to this idea